Streamlit UI Evaluation mode

PiffPaffM opened the pull request:
PR for issue #864

Timoeller commented on ui/webapp.py:
```diff
@@ -1,28 +1,105 @@
+
+# load csv into pandas dataframe
+if eval_mode:
+    df = pd.read_csv("eval_labels_example.csv", sep=";")
```
This we would need to configure from outside

PiffPaffM replied:
Yes,  there is a file with two example questions and answers. The user has to replace it with his data.

Timoeller replied:
K this hard coding of files would be an option. Cant we make it a parameter of our docker-compose file? I would prefer the second option if possible.

And what about the underlying documents? Currently we need to have them pre-indexed in ES. Can we either have a description on how to add the documents to an existing ES or add them in a spearate file like you did with questions?

Timoeller reviewed (commented):
I would love to see the functionality in our streamlit app.
I made a comment to make it more adjustable from the outside.

Does it make sense to create tests about this feature and resulting feedback labels?

Timoeller reviewed (commented):
Thanks for adding the detailed user feedback section to our docs.

I made some more comments on the code.

There was also an open [comment](https://github.com/deepset-ai/haystack/pull/920/files#r611731446) about file upload or attaching to an existing ES with pre indexed docs. Can we add some documentation on how to do that as well, please?

PiffPaffM commented on ui/README.md:
```diff
@@ -21,4 +26,14 @@ Just run
+
+The feedback can be exported with the API endpoint `export-doc-qa-feedback`. To learn more about finetuning a model with user feedback, please check out our [docs](https://haystack.deepset.ai/docs/latest/domain_adaptationmd#User-Feedback).
+
+![Screenshot](https://raw.githubusercontent.com/deepset-ai/haystack/master/docs/_src/img/streamlit_ui_screenshot_eval_mode.png)
```
The file has not been committed yet. It is located here: https://github.com/deepset-ai/haystack/raw/streamlit_ui_eval/docs/_src/img/streamlit_ui_screenshot_eval_mode.PNG

PiffPaffM commented on ui/st_state_patch.py:
```diff
@@ -0,0 +1,199 @@
+
+# Normally we'd use a Streamtit module, but I want a module that doesn't live in
+# your current working directory (since local modules get removed in between
+# runs), and Streamtit devs are likely to have Streamlit in their cwd.
```
Unfortunately, streamlit does not support any states out of the box. On every button click, streamlit reload the whole page and every value gets lost. This is kind of a downside of streamlit. To implement the intended behavior, we had to hack it a little bit. For more information please see [the changelog](https://docs.streamlit.io/en/stable/changelog.html?highlight=session#version-0-54-0). I guess there will be a better way to implement it in the future.

Timoeller replied:
Ok understood. This is not code you wrote but copy/pasted from "official" sources. Lets add a link to the source and make a comment about the reasoning for using this code. E.g. exactly what you just wrote: 

"streamlit does not support any states out of the box. On every button click, streamlit reload the whole page and every value gets lost." To keep track of our feedback state we use the official streamlit gist mentioned here...

In the linked changelog I also see another version of the "state hack". The st_state_patch doesnt seem to exist any longer? The newer [gist](https://gist.github.com/tvst/036da038ab3e999a64497f42de966a92) looks much more like code people can understand and build upon. Can we use that?

Timoeller reviewed (commented):
Some thoughts added on the st_state_patch.py

PiffPaffM commented:
@Timoeller I went through your comments and added more information. 
The gist [st_state_patch.py](https://gist.github.com/tvst/22ebc6d181a1b382c4d36dcee415f62f) still exists but the other implementation seems to be newer and is more likely the way going forward. Therefore, I changed that.

Timoeller commented:
Hey yes my bad. I saw these changes already briefly yesterday and was very happy about them. I did not find the time to properly check and approve the PR. Will do now.

Timoeller reviewed (approved):
Nice, thanks for adjusting the code.

About specifying `EVAL_FILE=eval_labels_example.csv` we can worry later. Because if somebody specifies a custom file that file still needs to somehow get into the docker at build time like you did with `COPY eval_labels_example.csv /home/user/`, correct?

PiffPaffM commented:
> Nice, thanks for adjusting the code.
> 
> About specifying `EVAL_FILE=eval_labels_example.csv` we can worry later. Because if somebody specifies a custom file that file still needs to somehow get into the docker at build time like you did with `COPY eval_labels_example.csv /home/user/`, correct?

Yes, that is true. Let's find a better way in the future

tholor commented:
> About specifying EVAL_FILE=eval_labels_example.csv we can worry later. Because if somebody specifies a custom file that file still needs to somehow get into the docker at build time like you did with COPY eval_labels_example.csv /home/user/, correct?

Why not mounting the file into the container at runtime? We could add a volume in the docker-compose ...
